core/Pos/Binary/Core/Address.hs

```diff
@@ -1,95 +1,189 @@
+        derivationPathListWithIndices =
+            case derivationPath of
+                Nothing -> []
+                -- 'unsafeFromJust' is safe, because 'case' ensures
+                -- that derivation path is 'Just'.
+                Just _ ->
+                    [(1, serialize' . unsafeFromJust . aaPkDerivationPath)]
+    decode = decodeAttributes initValue go
+      where
+        initValue =
+            AddrAttributes
+            { aaPkDerivationPath = Nothing
+            , aaStakeDistribution = BootstrapEraDistr
+            }
+        go n v acc =
+            case n of
+                0 -> pure acc {aaStakeDistribution = deserialize' v}
+                1 -> pure acc {aaPkDerivationPath = Just $ deserialize' v}
+                _ -> Nothing
+
+-- We don't need a special encoding for 'Address'', GND is what we want.
+deriving instance Bi Address'
+
+----------------------------------------------------------------------------
+-- Address serialization
+----------------------------------------------------------------------------
+
+{- NOTE: Address serialization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-----------------------------------------
+An address is serialized as a tuple consisting of:
 
-instance Bi (Attributes AddrPkAttrs) where
-    encode = encodeAttributes [(0, serialize' . addrPkDerivationPath)]
-    decode = decodeAttributes def $ \n v acc -> case n of
-        0 -> Just $ acc { addrPkDerivationPath = deserialize' v }
-        _ -> Nothing
+1. 'addrRoot'.
+2. 'addrAttributes'.
+3. 'addrType'.
+3. CRC32 checksum.
+-}
 
+instance CRC32 Address where
+    crc32Update seed =
+        crc32Update seed . CBOR.Write.toLazyByteString . encodeAddr
 
--- | Encodes the `Address` _without_ the CRC32.
+-- Encodes the `Address` __without__ the CRC32.
 -- It's important to keep this function separated from the `encode`
 -- definition to avoid that `encode` would call `crc32` and
 -- the latter invoke `crc32Update`, which would then try to call `encode`
 -- indirectly once again, in an infinite loop.
 encodeAddr :: Address -> Encoding
-encodeAddr addr = case addr of
-    PubKeyAddress keyHash attrs ->
-           encode (0 :: Word8)
-        <> encode (serialize' (keyHash, attrs))
-    ScriptAddress scrHash ->
-           encode (1 :: Word8)
-        <> encode (serialize' scrHash)
-    RedeemAddress keyHash ->
-           encode (2 :: Word8)
-        <> encode (serialize' keyHash)
-    UnknownAddressType t bs ->
-           encode t
-        <> encode bs
-
-instance Bi Address where
-    encode addr = encodeListLen 3 <> encodeAddr addr <> encode (crc32 addr)
+encodeAddr Address {..} =
+    encode addrRoot <> encode addrAttributes <> encode addrType
+
+-- Note: we are using 'Buildable' constraint here, which in turn
+-- relies on 'Bi', but it uses only encoding, while 'Buildable' is
+-- used here only in decoding.
+instance Buildable Address => Bi Address where
```

Comment:
but you don't use `Buildable` of `Address`, you use `Buildable` of CRC @gromakovsky